README.md

[![pub package](https://img.shields.io/pub/v/drop_cap_text.svg)](https://pub.dartlang.org/packages/drop_cap_text)
[![Awesome](https://img.shields.io/badge/Awesome-Flutter-blue.svg?longCache=true&style=flat-square)](https://stackoverflow.com/questions/tagged/flutter?sort=votes)


This Flutter plugin aims to achive [drop cap](https://en.wikipedia.org/wiki/Initial#Types_of_initial) in text capitalization.

## Usage

```dart
import 'package:drop_cap_text/drop_cap_text.dart';
```

```dart
DropCapText(
    loremIpsumText,
    style: TextStyle(fontStyle: FontStyle.italic),
),
```

![ex1](https://i.ibb.co/wQMn1z3/ex1.png)

### Properties

| Name                | Type               | Default                 | Description        |
| ------------------- | ------------------ | ----------------------- | ------------------ |
| data                | String             | null                    |                    |
| mode                | DropCapMode        | DropCapMode.inside      | aside, upwards, .. |
| textAlign           | TextAlign          | null                    |                    |
| indentation         | Offset             | Offset.zero             |                    |
| dropCapChars        | int                | 1                       |                    |
| dropCapPadding      | EdgeInsets         | EdgeInsets.zero         |                    |
| dropCap             | DropCap `[Widget]` | null                    |                    |
| style               | TextStyle          | null                    |                    |
| dropCapStyle        | TextStyle          |                         |                    |
| forceNoDescent      | bool               | false                   |                    |
| parseInlineMarkdown | bool               | false                   | bold|italic|underl |
| dropCapPosition     | DropCapPosition    | DropCapPosition.start   |                    |
| textDirection       | TextDirection      | TextDirection.ltr       |                    |
| maxLines            | int                |                         |                    |
| overflow            | TextOverflow       | TextOverflow.clip       |                    |



## Customization

##### Custom DropCap Widget: Image

![ex3](https://i.ibb.co/D43w1H8/ex3.png)

```dart
DropCapText(
    loremIpsumText,
    dropCap: DropCap(
    width: 100,
    height: 100,
    child: Image.network(
    	'https://www.codemate.com/wp-content/uploads/2017/09/flutter-logo.png')
    ),
),
```


------
##### Custom DropCap Widget: Parse Inline Markdown
only supports _italic_ **bold** underline

![ex8](https://imgbbb.com/images/2020/01/16/lorem_md.png)

```dart
DropCapText(
  'Lorem ipsum **dolor sit amet, consectetur adipiscing elit, ++sed do eiusmod++ tempor incididunt** ut labore et _dolore magna aliqua_.',
  parseInlineMarkdown: true,
  dropCapStyle: TextStyle(fontSize: 100, fontWeight: FontWeight.bold, color: Colors.green),
  dropCapPadding: EdgeInsets.only(right: 19.0),
  style: TextStyle(fontSize: 18.0, height: 1.5),
),
```


------
##### Custom DropCap Widget: Image right + justification

![ex7](https://i.ibb.co/WVPT3HH/img-end.jpg)

```dart
DropCapText(
    loremIpsumText,
    dropCapPosition: DropCapPosition.end,
    textAlign: TextAlign.justify,
    dropCap: DropCap(
    width: 100,
    height: 100,
    child: Image.network(
    	'https://www.codemate.com/wp-content/uploads/2017/09/flutter-logo.png')
    ),
),
```

------

##### 2 characters + indentation 

![ex2](https://i.ibb.co/yq1Vj7q/ex2.png)

```dart
DropCapText(
    loremIpsumText,
    style: TextStyle(
        height: 1.3,
        fontFamily: 'times',
    ),
    dropCapChars: 2,
    indentation: Offset(25, 10),
),
```

------

##### Upward drop cap 

![ex5](https://i.ibb.co/b3M6KD8/ex5.png)

```dart
DropCapText(
    loremIpsumText,
    mode: DropCapMode.upwards,
    dropCapStyle: TextStyle(
        color: Colors.red,
        fontWeight: FontWeight.bold,
        fontSize: 120,
        fontFamily: 'times',
    ),
),
```

------

##### Aside drop cap

![ex6](https://i.ibb.co/bFmrM6G/ex6.png)

```dart
DropCapText(
    loremIpsumText,
    style: TextStyle(
        fontWeight: FontWeight.bold,
        height: 1.2,
    ),
    mode: DropCapMode.aside,
),
```



CHANGELOG.md
### TODO

* support advanced markdown 


lib/drop_cap_text.dart
library drop_cap_text;

import 'dart:math';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

enum DropCapMode {
  /// default
  inside,
  upwards,
  aside,

  /// Does not support dropCapPadding, indentation, dropCapPosition and custom dropCap.
  /// Try using DropCapMode.upwards in combination with dropCapPadding and forceNoDescent=true
  baseline
}

enum DropCapPosition {
  start,
  end,
}

class DropCap extends StatelessWidget {
  final Widget child;
  final double width, height;

  DropCap({
    Key key,
    this.child,
    @required this.width,
    @required this.height,
  })  : assert(width != null),
        assert(height != null),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    return SizedBox(child: child, width: width, height: height);
  }
}

class DropCapText extends StatelessWidget {
  final String data;
  final DropCapMode mode;
  final TextStyle style, dropCapStyle;
  final TextAlign textAlign;
  final DropCap dropCap;
  final EdgeInsets dropCapPadding;
  final Offset indentation;
  final bool forceNoDescent, parseInlineMarkdown;
  final TextDirection textDirection;
  final DropCapPosition dropCapPosition;
  final int dropCapChars;
  final int maxLines;
  final TextOverflow overflow;

  DropCapText(this.data,
      {Key key,
      this.mode = DropCapMode.inside,
      this.style,
      this.dropCapStyle,
      this.textAlign = TextAlign.start,
      this.dropCap,
      this.dropCapPadding = EdgeInsets.zero,
      this.indentation = Offset.zero,
      this.dropCapChars = 1,
      this.forceNoDescent = false,
      this.parseInlineMarkdown = false,
      this.textDirection = TextDirection.ltr,
      this.overflow = TextOverflow.clip,
      this.maxLines,
      this.dropCapPosition})
      : assert(data != null),
        super(key: key);

  @override
  Widget build(BuildContext context) {
    TextStyle textStyle = TextStyle(
      color: Colors.black,
      fontSize: 14,
      height: 1,
      fontFamily: Theme.of(context)?.textTheme?.body1?.fontFamily,
    ).merge(style);

    if (data == '') return Text(data, style: textStyle);

    TextStyle capStyle = TextStyle(
      color: textStyle.color,
      fontSize: textStyle.fontSize * 5.5,
      fontFamily: textStyle.fontFamily,
      fontWeight: textStyle.fontWeight,
      fontStyle: textStyle.fontStyle,
      height: 1,
    ).merge(dropCapStyle);

    double capWidth, capHeight;
    int dropCapChars = dropCap != null ? 0 : this.dropCapChars;
    CrossAxisAlignment sideCrossAxisAlignment = CrossAxisAlignment.start;
    MarkdownParser mdData = parseInlineMarkdown ? MarkdownParser(data) : null;
    final String dropCapStr = (mdData?.plainText ?? data).substring(0, dropCapChars);

    if (mode == DropCapMode.baseline && dropCap == null) return _buildBaseline(context, textStyle, capStyle);

    // custom DropCap
    if (dropCap != null) {
      capWidth = dropCap.width;
      capHeight = dropCap.height;
    } else {
      TextPainter capPainter = TextPainter(
        text: TextSpan(
          text: dropCapStr,
          style: capStyle,
        ),
        textDirection: textDirection,
      );
      capPainter.layout();
      capWidth = capPainter.width;
      capHeight = capPainter.height;
      if (forceNoDescent) {
        List<LineMetrics> ls = capPainter.computeLineMetrics();
        capHeight -= ls.isNotEmpty ? ls[0].descent * 0.95 : capPainter.height * 0.2;
      }
    }

    // compute drop cap padding
    capWidth += dropCapPadding.left + dropCapPadding.right;
    capHeight += dropCapPadding.top + dropCapPadding.bottom;

    MarkdownParser mdRest = parseInlineMarkdown ? mdData.subchars(dropCapChars) : null;
    String restData = data.substring(dropCapChars);

    TextSpan textSpan = TextSpan(
      text: parseInlineMarkdown ? null : restData,
      children: parseInlineMarkdown ? mdRest.toTextSpanList() : null,
      style: textStyle.apply(fontSizeFactor: MediaQuery.of(context).textScaleFactor),
    );

    TextPainter textPainter = TextPainter(textDirection: textDirection, text: textSpan, textAlign: textAlign);
    double lineHeight = textPainter.preferredLineHeight;

    int rows = ((capHeight - indentation.dy) / lineHeight).ceil();

    // DROP CAP MODE - UPWARDS
    if (mode == DropCapMode.upwards) {
      rows = 1;
      sideCrossAxisAlignment = CrossAxisAlignment.end;
    }

    // BUILDER
    return LayoutBuilder(builder: (BuildContext context, BoxConstraints constraints) {
      double boundsWidth = constraints.maxWidth - capWidth;
      if (boundsWidth < 1) boundsWidth = 1;

      int charIndexEnd = data.length;

      //int startMillis = new DateTime.now().millisecondsSinceEpoch;
      if (rows > 0) {
        textPainter.layout(maxWidth: boundsWidth);
        double yPos = rows * lineHeight;
        int charIndex = textPainter.getPositionForOffset(Offset(0, yPos)).offset;
        textPainter.maxLines = rows;
        textPainter.layout(maxWidth: boundsWidth);
        if (textPainter.didExceedMaxLines) charIndexEnd = charIndex;
      } else {
        charIndexEnd = dropCapChars;
      }
      //int totMillis = new DateTime.now().millisecondsSinceEpoch - startMillis;

      // DROP CAP MODE - LEFT
      if (mode == DropCapMode.aside) charIndexEnd = data.length;

      return Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          // Text(totMillis.toString() + ' ms'),
          Row(
            textDirection: dropCapPosition == null || dropCapPosition == DropCapPosition.start
                ? textDirection
                : (textDirection == TextDirection.ltr ? TextDirection.rtl : TextDirection.ltr),
            crossAxisAlignment: sideCrossAxisAlignment,
            children: <Widget>[
              dropCap != null
                  ? Padding(padding: dropCapPadding, child: dropCap)
                  : Container(
                      width: capWidth,
                      height: capHeight,
                      padding: dropCapPadding,
                      child: RichText(
                        textDirection: textDirection,
                        textAlign: textAlign,
                        text: TextSpan(
                          text: dropCapStr,
                          style: capStyle,
                        ),
                      ),
                    ),
              Flexible(
                child: Container(
                  padding: EdgeInsets.only(top: indentation.dy),
                  height: (lineHeight * min(maxLines ?? rows, rows)) + indentation.dy,
                  child: RichText(
                    overflow: (maxLines == null || (maxLines > rows && overflow == TextOverflow.fade))
                        ? TextOverflow.clip
                        : overflow,
                    maxLines: maxLines,
                    textDirection: textDirection,
                    textAlign: textAlign,
                    text: textSpan,
                  ),
                ),
              ),
            ],
          ),
          if (maxLines == null || maxLines > rows)
            Padding(
              padding: EdgeInsets.only(left: indentation.dx),
              child: RichText(
                overflow: overflow,
                maxLines: maxLines != null && maxLines > rows ? maxLines - rows : null,
                textAlign: textAlign,
                textDirection: textDirection,
                text: TextSpan(
                  text: parseInlineMarkdown ? null : restData.substring(charIndexEnd),
                  children: parseInlineMarkdown ? mdRest.subchars(charIndexEnd).toTextSpanList() : null,
                  style: textStyle.apply(fontSizeFactor: MediaQuery.of(context).textScaleFactor),
                ),
              ),
            ),
        ],
      );
    });
  }

  _buildBaseline(BuildContext context, TextStyle textStyle, TextStyle capStyle) {
    MarkdownParser mdData = MarkdownParser(data);

    return RichText(
      textAlign: textAlign,
      text: TextSpan(
        style: textStyle,
        children: <TextSpan>[
          TextSpan(
            text: mdData.plainText.substring(0, dropCapChars),
            style: capStyle.merge(TextStyle(height: 0)),
          ),
          TextSpan(
            children: mdData.subchars(dropCapChars).toTextSpanList(),
            style: textStyle.apply(fontSizeFactor: MediaQuery.of(context).textScaleFactor),
          ),
        ],
      ),
    );
  }
}

class MarkdownParser {
  final String data;
  List<MarkdownSpan> spans;
  String plainText;

  List<TextSpan> toTextSpanList() {
    return spans.map((s) => s.toTextSpan()).toList();
  }

  MarkdownParser subchars(int startIndex, [int endIndex]) {
    final List<MarkdownSpan> subspans = [];
    int skip = startIndex;
    for (int s = 0; s < spans.length; s++) {
      MarkdownSpan span = spans[s];
      if (skip <= 0) {
        subspans.add(span);
      } else if (span.text.length < skip) {
        skip -= span.text.length;
      } else {
        subspans.add(
          MarkdownSpan(
            style: span.style,
            markups: span.markups,
            text: span.text.substring(skip, span.text.length),
          ),
        );
        skip = 0;
      }
    }

    return MarkdownParser(
      subspans
          .asMap()
          .map((int index, MarkdownSpan span) {
            String markup = index > 0
                ? (span.markups.isNotEmpty ? span.markups[0].code : '')
                : span.markups.map((m) => m.isActive ? m.code : '').join();
            return MapEntry(index, '$markup${span.text}');
          })
          .values
          .toList()
          .join(),
    );
  }

  MarkdownParser(this.data) {
    plainText = '';
    spans = [MarkdownSpan(text: '', markups: [], style: TextStyle())];

    bool bold = false;
    bool italic = false;
    bool underline = false;

    const String MARKUP_BOLD = '**';
    const String MARKUP_ITALIC = '_';
    const String MARKUP_UNDERLINE = '++';

    addSpan(String markup, bool isOpening) {
      final List<Markup> markups = [Markup(markup, isOpening)];

      if (bold && markup != MARKUP_BOLD) markups.add(Markup(MARKUP_BOLD, true));
      if (italic && markup != MARKUP_ITALIC) markups.add(Markup(MARKUP_ITALIC, true));
      if (underline && markup != MARKUP_UNDERLINE) markups.add(Markup(MARKUP_UNDERLINE, true));

      spans.add(
        MarkdownSpan(
          text: '',
          markups: markups,
          style: TextStyle(
            fontWeight: bold ? FontWeight.bold : null,
            fontStyle: italic ? FontStyle.italic : null,
            decoration: underline ? TextDecoration.underline : null,
          ),
        ),
      );
    }

    bool checkMarkup(int i, String markup) {
      return data.substring(i, min(i + markup.length, data.length)) == markup;
    }

    for (int c = 0; c < data.length; c++) {
      if (checkMarkup(c, MARKUP_BOLD)) {
        bold = !bold;
        addSpan(MARKUP_BOLD, bold);
        c += MARKUP_BOLD.length - 1;
      } else if (checkMarkup(c, MARKUP_ITALIC)) {
        italic = !italic;
        addSpan(MARKUP_ITALIC, italic);
        c += MARKUP_ITALIC.length - 1;
      } else if (checkMarkup(c, MARKUP_UNDERLINE)) {
        underline = !underline;
        addSpan(MARKUP_UNDERLINE, underline);
        c += MARKUP_UNDERLINE.length - 1;
      } else {
        spans[spans.length - 1].text += data[c];
        plainText += data[c];
      }
    }
  }
}

class MarkdownSpan {
  final TextStyle style;
  final List<Markup> markups;
  String text;

  TextSpan toTextSpan() => TextSpan(text: text, style: style);

  MarkdownSpan({
    @required this.text,
    @required this.style,
    @required this.markups,
  });
}

class Markup {
  final String code;
  final bool isActive;

  Markup(this.code, this.isActive);
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:drop_cap_text/drop_cap_text.dart';

final String loremIpsumText =
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum";

void main() {
  runApp(new ExampleApp());
}

class ExampleApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: Scaffold(
        body: SingleChildScrollView(
          child: Padding(
            padding: EdgeInsets.all(20),
            child: Column(
              children: <Widget>[
                DropCapText(
                  loremIpsumText,
                  style: TextStyle(fontStyle: FontStyle.italic),
                ),
                SizedBox(height: 30),
                DropCapText(
                  loremIpsumText,
                  dropCap: DropCap(
                      width: 100,
                      height: 100,
                      child: Image.network('https://www.codemate.com/wp-content/uploads/2017/09/flutter-logo.png')),
                ),
                SizedBox(height: 30),
                DropCapText(
                  'Lorem ipsum **dolor sit amet, consectetur adipiscing elit, ++sed do eiusmod++ tempor incididunt** ut labore et _dolore magna aliqua_.',
                  parseInlineMarkdown: true,
                  dropCapStyle: TextStyle(fontSize: 100, fontWeight: FontWeight.bold, color: Colors.green),
                  dropCapPadding: EdgeInsets.only(right: 19.0),
                  style: TextStyle(fontSize: 18.0, height: 1.5),
                ),
                SizedBox(height: 30),
                DropCapText(
                  loremIpsumText,
                  dropCapPosition: DropCapPosition.end,
                  textAlign: TextAlign.justify,
                  dropCap: DropCap(
                      width: 100,
                      height: 100,
                      child: Image.network('https://www.codemate.com/wp-content/uploads/2017/09/flutter-logo.png')),
                ),
                SizedBox(height: 30),
                DropCapText(
                  loremIpsumText,
                  style: TextStyle(
                    height: 1.3,
                    fontFamily: 'times',
                  ),
                  dropCapChars: 2,
                  indentation: Offset(25, 10),
                ),
                SizedBox(height: 30),
                DropCapText(
                  loremIpsumText,
                  mode: DropCapMode.upwards,
                  dropCapStyle: TextStyle(
                    color: Colors.red,
                    fontWeight: FontWeight.bold,
                    fontSize: 120,
                    fontFamily: 'times',
                  ),
                ),
                SizedBox(height: 30),
                DropCapText(
                  loremIpsumText,
                  style: TextStyle(
                    fontWeight: FontWeight.bold,
                    height: 1.2,
                  ),
                  mode: DropCapMode.aside,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
